### ISSUE TITLE
Weird triangle artifact present in many images

### ISSUE THREAD

**Issue by xiankgx** [jingyuanli001/RFR-Inpainting#39]
Dear @jingyuanli001 , hi, I've been training your model with the places2 large dataset for quite some time, over few million iterations, and I've been seeing the same weird triangle artifact appearing in many images.

Some modifications I've done are:
- Increasing image training size to: 480
- Increasing hole size by multiplying max_width and max_length both by 1.5 in the random hole generation function.
- Increasing the RFR iterations or recurrence from 6 to 13 to account for the larger image and holes.

I think these artifacts appear when the holes are bigger. Have you encountered this problem? Please help, thank you.

![all_seeing_eye](https://user-images.githubusercontent.com/4113258/109731848-0078fe00-7bf7-11eb-9e2b-ca3bf80fbd81.png)
![all_seeing_eye2](https://user-images.githubusercontent.com/4113258/109731857-040c8500-7bf7-11eb-8410-6fcaad713304.png)
![all_seeing_eye3](https://user-images.githubusercontent.com/4113258/109731865-066edf00-7bf7-11eb-9b38-bb94b5715d24.png)

**Comment by jingyuanli001** [comment]
Hi, thanks for your question.
In fact, this is also an issue we observed when developing the method (as well as testing a few other inpainting methods). Sometimes this issue is more significant. From my personal experience, there are three directions to work on to address the issue. First, you might try to tune the hyper-parameters with respect to perceptual loss and style loss, which might be helpful for addressing the artifact. Second, you might try to replace batch norm with some other normalization method to address this issue. Third, add an adversarial loss. This would require extra computation but can address the artifact a lot more effectively). We are not quite sure about the nature of this artifact though, but obviously by addressing these artifacts the evaluation results would be even better.

**Comment by jingyuanli001** [comment]
This issue seems to be more significant when the masked region is large. It's possible that, because most of the masks used during training is not that large as some extreme cases, the model prefers to give good results on most of the training cases and ignores some extreme cases that are less common during training.

**Comment by xiankgx** [comment]
Thanks for your swift reply. 

But this artifact is not during testing, but also during training. Like I said, I've increased the difficulty of the training by increasing the max stroke width and length by a constant multiplier.

Also, this artifact is more evident the longer we train. It is not very noticeable during the earlier iterations.

Somehow, the very same artifact appearing in many images is disturbing. It would be better if the artifacts look different.

Also, it somehow gives me the feeling of hitting some "local minimum". It is like you have a bubble between a soft sticker and a hard surface (mirror) and we try to remove the bubble by trying to move it out by pressing with our finger. It is possible it could appear that the trapped air can't go anywhere and further trying to fix the problem resulted in these "peaks".

**Comment by xiankgx** [comment]
I hope this explain what I felt with my last post.

![Removing-Bubbles](https://user-images.githubusercontent.com/4113258/110084230-ae88d180-7dca-11eb-88f5-c1962b30cdfb.jpg)

**Comment by xiankgx** [comment]
Anyway, thanks for the swift reply. We will try our your suggestions.

**Comment by jingyuanli001** [comment]
> Thanks for your swift reply.
> 
> But this artifact is not during testing, but also during training. Like I said, I've increased the difficulty of the training by increasing the max stroke width and length by a constant multiplier.
> 
> Also, this artifact is more evident the longer we train. It is not very noticeable during the earlier iterations.
> 
> Somehow, the very same artifact appearing in many images is disturbing. It would be better if the artifacts look different.
> 
> Also, it somehow gives me the feeling of hitting some "local minimum". It is like you have a bubble between a soft sticker and a hard surface (mirror) and we try to remove the bubble by trying to move it out by pressing with our finger. It is possible it could appear that the trapped air can't go anywhere and further trying to fix the problem resulted in these "peaks".

That makes sense. But I am also feeling that, if you input the reconstructed images with artifacts into the model and re-inpaint the region with artifact, the artifact won't exist. Would this indicate that the last few iterations in RFR are actually less efficient, due to the overly large amount of information to process and parse?

**Comment by xiankgx** [comment]
Yeah, I agree that it would be possible to just detect these artifacts and redo a second round and have no problem with it.

Perhaps this is a result of iterative processing, when you reach the end, it is **difficult to fix things without readjusting a major part of the other pixels**. Just what I feel.

Please let us know if you have come across a solution. Anyway, I think the idea you presented in your paper is really novel and stands out from the rest. The result is really good too other than this artifact problem.

**Comment by xiankgx** [comment]
In other words, the errors are concentrated at the pixels in the final iteration, rather than evenly distributed across all the inpainted pixels.

**Comment by jingyuanli001** [comment]
> Yeah, I agree that it would be possible to just detect these artifacts and redo a second round and have no problem with it.
> 
> Perhaps this is a result of iterative processing, when you reach the end, it is **difficult to fix things without readjusting a major part of the other pixels**. Just what I feel.
> 
> Please let us know if you have come across a solution. Anyway, I think the idea you presented in your paper is really novel and stands out from the rest. The result is really good too other than this artifact problem.

Thanks! I will get back to you if we have more information.

**Comment by nachifur** [comment]
@jingyuanli001  Thank you very much for sharing your awesome work. I am wondering if you plan to share the Places2 weight?
